// @orxe-culture/core/dist/packages/orxe-culture-core/src/core/config/index.js
import { defaultInterpolationFormat } from '../formats';
import { CultureDefaultConfig } from './culture-default-config';
export class CultureConfig {
    constructor(data) {
        this.ns = data && data.ns ? data.ns : CultureDefaultConfig.ns;
        this.lng = data && data.lng ? data.lng : CultureDefaultConfig.lng;
        this.fallbackLng =
            data && data.fallbackLng ? data.fallbackLng : CultureDefaultConfig.fallbackLng;
        this.debug = data && data.debug ? data.debug : CultureDefaultConfig.debug;
        this.saveMissing =
            data && data.saveMissing ? data.saveMissing : CultureDefaultConfig.saveMissing;
        this.production = data && data.production ? data.production : CultureDefaultConfig.production;
        this.isInsideMicroFrontEnd =
            data && data.isInsideMicroFrontEnd
                ? data.isInsideMicroFrontEnd
                : CultureDefaultConfig.isInsideMicroFrontEnd;
        this.interpolation = {
            format: defaultInterpolationFormat,
        };
        this.fallbackNS = data && data.fallbackNS ? data.fallbackNS : CultureDefaultConfig.fallbackNS;
        this.resources = data && data.resources ? data.resources : CultureDefaultConfig.resources;
        this.detection = CultureDefaultConfig.detection;
    }
}
