// @orxe-culture/core/dist/packages/orxe-culture-core/src/core/formats/number-format/index.js
/* eslint-disable @typescript-eslint/ban-ts-ignore */
import Currency from './currency';
// Following Number types can be supported with help of INTL API for Numbers
// We need currency and decimal format for now please add more functions if any new functionality needs to be supported
// Refer: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat
export const NumberFormatPartTypes = {
    CURRENCY: 'currency',
    DECIMAL: 'decimal',
    FRACTION: 'fraction',
    GROUP: 'group',
    INFINITY: 'infinity',
    INTEGER: 'integer',
    LITERAL: 'literal',
};
export const decimal = (number, lang) => {
    return new Intl.NumberFormat(lang, { style: NumberFormatPartTypes.DECIMAL }).format(number);
};
export const currency = (number, lang) => {
    // FallBack to USD
    return new Intl.NumberFormat(lang, {
        style: NumberFormatPartTypes.CURRENCY,
        currency: Currency[lang] ? Currency[lang] : Currency['en-US'],
        useGrouping: true,
    }).format(number);
};
