// @orxe-culture/core/dist/core/methods/index.js
import { i18nDefault } from '../';
import { CultureDefaultConfig } from '../config/culture-default-config';
/**
 *
 * Instead of exposing complete i18next instance to the user which they might reconfigure,exposing translate function
 * Translate is a reserve word in HTML-ELEMENT class so exposing this function by name t
 */
export const t = i18nDefault.t.bind(i18nDefault);
export const format = (value, format, lng) => {
    return i18nDefault.format.call(i18nDefault, value, format, lng);
};
export const changeLanguage = (lng) => {
    return i18nDefault.changeLanguage.call(i18nDefault, lng);
};
/**
 *
 *
 * As per the requirements resources would be fetched and added to i18Next on runtime.
 * Documentation of these method is available on https://www.i18next.com/overview/api
 * Extracting this method in Culture class to force deep and overwrite option.
 * Default NameSpace is translation, and default lang is 'en-US'.
 */
export const addResourceBundle = (resourceConfig) => {
    const { lng = CultureDefaultConfig.lng, ns = CultureDefaultConfig.fallbackNS, resources, } = resourceConfig;
    return i18nDefault.addResourceBundle.call(i18nDefault, lng, ns, resources, false, true);
};
