// @orxe-components/button/dist/button-css.js
import { css } from 'lit-element';
export default css `
  :host {
    --button-font-size-small: 1.4rem;
    --button-height-small: 36px;
    --button-width: 100%;
    --icon-touch-target: var(--touch-target-large);
    --button-padding-1: 0 var(--spacing-03);
    --button-padding-2: 0;
    --button-padding-3: 0;
    --button-tertiary-border-size: 2px;
    --button-font-family: var(--global-font-family);
    --button-border-radius: var(--border-radius);
    --button-font-weight: var(--font-weight-semi-bold);
    --button-primary-background-color: var(--primary);
    --button-primary-font-color: var(--neutral);
    --button-secondary-background-color: var(--secondary);
    --button-secondary-font-color: var(--neutral);
    --button-tertiary-background-color: var(--neutral);
    --button-tertiary-font-color: var(--secondary);
    --button-icon-color: currentColor;
    --button-icon-withlabel-icon-color: var(--secondary);
    --button-icon-withlabel-font-color: var(--secondary);
    --button-floating-background-color: var(--neutral);
    --button-floating-box-shadow: 0 8px 12px 3px rgba(0, 0, 0, 0.16);
    --button-floating-font-color: var(--secondary);
    --button-floating-border-radius: 4px;
    --button-link-font-color: var(--links);
    --button-touch-target-small: var(--touch-target-small);
    --button-touch-target-medium: var(--touch-target-medium);
    --button-touch-target-large: var(--touch-target-large);
    --link-touch-target: var(--touch-target-small);
    --float-button-touch-targe: var(--touch-target-medium);
    --icon-and-text-touch-target: var(--touch-target-small);
    --icon-touch-target: var(--touch-target-large);
  }
  :host .button-container {
    display: unset;
  }
  :host .orxe-button {
    border-width: 0;
    position: relative;
    overflow: hidden;
    font-family: var(--button-font-family);
    font-weight: var(--button-font-weight);
    cursor: pointer;
  }
  :host .orxe-button slot {
    pointer-events: none;
  }
  :host .orxe-button:disabled {
    cursor: default;
  }
  :host .orxe-btn--extra-small {
    font-size: 1.2rem;
    height: 24px;
    min-width: 48px;
    max-width: 100%;
    padding: var(--button-padding-1);
  }
  :host .orxe-btn--small {
    font-size: var(--button-font-size-small);
    min-height: var(--button-touch-target-small);
    min-width: 80px;
    max-width: 100%;
    width: var(--button-width);
    padding: var(--button-padding-2);
  }
  :host .orxe-btn--medium {
    font-size: 1.6rem;
    min-height: var(--button-touch-target-medium);
    width: var(--button-width);
    max-width: 100%;
    min-width: 80px;
    padding: var(--button-padding-3);
  }
  :host .orxe-btn--large {
    font-size: 1.6rem;
    min-height: var(--button-touch-target-large);
    max-width: 100%;
    width: var(--button-width);
    min-width: 80px;
    padding: var(--button-padding-3);
  }
  :host .orxe-link {
    display: inline-flex;
    align-items: center;
    max-width: 100%;
    padding: 0 var(--spacing-03);
    font-weight: var(--button-font-weight);
    color: var(--button-link-font-color);
    min-height: var(--link-touch-target);
    position: relative;
    overflow: hidden;
    cursor: pointer;
  }
  :host .orxe-link slot {
    pointer-events: none;
  }
  :host .orxe-link--large {
    font-size: 1.4rem;
  }
  :host .orxe-link--small {
    font-size: 1.2rem;
  }
  :host .primary {
    background-color: var(--button-primary-background-color);
    color: var(--button-primary-font-color);
    border-radius: var(--button-border-radius);
    border: 2px solid var(--button-primary-background-color);
  }
  :host .primary:disabled {
    border: 2px solid var(--disabled-text);
    background-color: var(--disabled-text);
    color: var(--button-primary-font-color);
    box-shadow: none;
  }
  :host .primary .ripple-effect {
    background-color: var(--neutral);
  }
  :host .secondary {
    background-color: var(--button-secondary-background-color);
    color: var(--button-secondary-font-color);
    border-radius: var(--button-border-radius);
    border: 2px solid var(--button-secondary-background-color);
  }
  :host .secondary:disabled {
    border: 2px solid var(--disabled-text);
    background-color: var(--disabled-text);
    color: var(--button-secondary-font-color);
    box-shadow: none;
  }
  :host .secondary .ripple-effect {
    background-color: var(--neutral);
  }
  :host .tertiary {
    background-color: var(--button-tertiary-background-color);
    color: var(--button-tertiary-font-color);
    border: var(--button-tertiary-border-size) solid var(--button-tertiary-font-color);
    border-radius: var(--button-border-radius);
  }
  :host .tertiary:disabled {
    background-color: var(--button-tertiary-background-color);
    color: var(--disabled-text);
    border: var(--button-tertiary-border-size) solid var(--disabled-text);
  }
  :host .tertiary .ripple-effect {
    background-color: var(--separator-01);
  }
  :host .iconAndText {
    cursor: pointer;
    max-width: 100%;
    min-height: var(--icon-and-text-touch-target);
    display: inline-flex;
    align-items: center;
    padding: 0 var(--spacing-03);
    position: relative;
    overflow: hidden;
    text-decoration: none;
  }
  :host .iconAndText slot {
    pointer-events: none;
    font-size: 1.4rem;
    font-weight: var(--button-font-weight);
    color: var(--button-icon-withlabel-font-color);
    font-family: var(--button-font-family);
  }
  :host .iconAndText orxe-icon {
    pointer-events: none;
    padding-right: var(--spacing-02);
    color: var(--button-icon-withlabel-icon-color);
  }
  :host .float {
    width: 100%;
    height: 36px;
    min-width: 80px;
    max-width: 100%;
    background-color: transparent;
    font-size: 1.6rem;
    color: var(--button-floating-font-color);
    display: inline-flex;
    align-items: center;
    justify-content: center;
    border: 0;
    border-radius: 0;
    padding: 0 var(--spacing-05);
    cursor: pointer;
    -webkit-tap-highlight-color: transparent;
  }
  :host .iconOnly {
    min-width: var(--icon-touch-target);
    min-height: var(--icon-touch-target);
    display: inline-flex;
    align-items: center;
    justify-content: center;
    background-color: transparent;
  }
  :host .iconOnly svg {
    pointer-events: none;
  }
  :host .iconOnly svg path {
    fill: var(--button-icon-color);
  }
  :host-context([kind='float']) .button-container {
    top: -2px;
    align-items: center;
    height: 40px;
    display: flex;
  }
  :host-context([kind='float']) .ripple-effect {
    top: calc(var(--float-btn-ripple-top) - (100vh - 100% - 1rem)) !important;
    left: calc(var(--float-btn-ripple-left) - (50vw - 100%)) !important;
    transform: translate(-50%, -50%);
  }
  :host([kind='iconOnly']) orxe-icon {
    color: var(--button-icon-color);
    pointer-events: none;
  }
  :host([kind='iconOnly']) .iconOnly {
    display: flex;
    align-items: center;
  }
  :host([kind='iconOnly']) .ripple-effect {
    top: 50% !important;
    left: 50% !important;
    transform: translate(-50%, -50%);
  }
  :host([href]) {
    line-height: 1;
  }
  :host([size='extra-small']) .button-container {
    display: inline-flex;
    align-items: center;
    min-height: var(--button-touch-target-medium);
  }
  :host([kind='iconAndText'][icon-size='medium']) slot {
    font-size: var(--button-icon-and-text-font-size-medium);
  }
  :host([kind='iconAndText'][icon-size='small']) slot {
    font-size: var(--button-icon-and-text-font-size-small);
  }
`;
